### PR TITLE
Remove SpawnModules

### LINKED ISSUE #12783: Remove SpawnModules
This is a spinoff from #12744. SpawnModules is a way to have modules produce sub modules. However, one way they have been used in the past is broken since we now have plugin classloader isolation. Some extension points used to allow plugging in an entire module "implementation". This is far too crazy for us to maintain in ES. It also leads to having lots of tiny modules that bind just a handful of classes.

We should remove this interface, and flatten the modules we have. All extension points should be class implementation based, by registering custom implementations with the module responsible for binding a given class. This has a bunch of benefits, like:
* Making it easier to find where to add new classes to bind
* Making it easier to document and write plugins since there will be less modules to possibly register custom extensions with
* Reducing spaghetti dependencies because it is clear when we have *tons* of bound classes (a lot of which can probably be removed when we can clearly see what they are instead of being spread across tons of modules)

### PR DESCRIPTION
**Pull request by rjernst**
The last use case of spawn modules was for plugins. This change handles
plugin modules directly, and removes SpawnModules.

closes #12783

### CODE REVIEW

**Comment by rjernst** [comment]
I synced up with master, and also removed PreProcessModule as it is no longer used.

File: core/src/main/java/org/elasticsearch/indices/IndicesService.java [hunk 1]
```diff
@@ -316,6 +333,11 @@ public synchronized IndexService createIndex(String sIndexName, @IndexSettings S
         modules.add(new IndexAliasesServiceModule());
         modules.add(new IndexModule(indexSettings));
 
+        for (Module pluginModule : pluginsService.indexModules(indexSettings)) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/indices/IndicesService.java]: is this actually used anywhere? I mean why would you add your own index modules? I really wonder if we need this kind of extension mechanism? I know it's unrelated but maybe we need a followup issue?

**Comment by s1monw** [comment]
LGTM

File: core/src/main/java/org/elasticsearch/indices/IndicesService.java (hunk 1 above)
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/indices/IndicesService.java]: I created #13059

**Comment by clintongormley** [comment]
@rjernst Did you mean to merge this in to 2.0 as well?

**Comment by s1monw** [comment]
back-ported to 2.0 branch as well

**Pull request merged**